From: Tom Herbert <therbert@google.com>
Date: Mon, 28 Nov 2011 16:33:30 +0000
Subject: tg3: Support for byte queue limits

commit 298376d3e8f00147548c426959ce79efc47b669a upstream.

Changes to tg3 to use byte queue limits.

Signed-off-by: Tom Herbert <therbert@google.com>
Acked-by: Eric Dumazet <eric.dumazet@gmail.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
[bwh: Adjust to apply after backported commit 497a27b9e1bc
 'tg3: Call dev_kfree_skby_any instead of dev_kfree_skb.']
---
 drivers/net/ethernet/broadcom/tg3.c |    8 ++++++++
 1 file changed, 8 insertions(+)

--- a/drivers/net/ethernet/broadcom/tg3.c
+++ b/drivers/net/ethernet/broadcom/tg3.c
@@ -5362,6 +5362,7 @@ static void tg3_tx(struct tg3_napi *tnap
 	u32 sw_idx = tnapi->tx_cons;
 	struct netdev_queue *txq;
 	int index = tnapi - tp->napi;
+	unsigned int pkts_compl = 0, bytes_compl = 0;
 
 	if (tg3_flag(tp, ENABLE_TSS))
 		index--;
@@ -5412,6 +5413,9 @@ static void tg3_tx(struct tg3_napi *tnap
 			sw_idx = NEXT_TX(sw_idx);
 		}
 
+		pkts_compl++;
+		bytes_compl += skb->len;
+
 		dev_kfree_skb_any(skb);
 
 		if (unlikely(tx_bug)) {
@@ -5420,6 +5424,8 @@ static void tg3_tx(struct tg3_napi *tnap
 		}
 	}
 
+	netdev_completed_queue(tp->dev, pkts_compl, bytes_compl);
+
 	tnapi->tx_cons = sw_idx;
 
 	/* Need to make the tx_cons update visible to tg3_start_xmit()
@@ -6863,6 +6869,7 @@ static netdev_tx_t tg3_start_xmit(struct
 	}
 
 	skb_tx_timestamp(skb);
+	netdev_sent_queue(tp->dev, skb->len);
 
 	/* Packets are ready, update Tx producer idx local and on card. */
 	tw32_tx_mbox(tnapi->prodmbox, entry);
@@ -7344,6 +7351,7 @@ static void tg3_free_rings(struct tg3 *t
 			dev_kfree_skb_any(skb);
 		}
 	}
+	netdev_reset_queue(tp->dev);
 }
 
 /* Initialize tx/rx rings for packet processing.
